platform/lib/build/pageTransformer.js

```diff
@@ -116,14 +120,123 @@ class PageMinifier {
           canonicalPage.contents = Buffer.from(optimizedHtml);
           ampPage.contents = Buffer.from(html);
 
-          this.push(canonicalPage);
-          this.push(ampPage);
+          let filteredPages = [];
+          if (isFilterableRoute(canonicalPage.path)) {
+            filteredPages = scope._filterPages(canonicalPage, ampPage);
+          }
 
+          for (const page of [canonicalPage, ampPage, ...filteredPages]) {
+            this.push(page);
+          }
+
+          scope._log.success(`Transformed ${canonicalPage.path}`);
           callback();
         }))
-        .pipe(gulp.dest('./'));
+        .pipe(gulp.dest(project.paths.PAGES_DEST));
+  }
+
+
+  /**
+   * Verifies a given page should be filtered
+   *
+   * @param  {Vinyl}
+   * @return {undefined}
+   */
+  _isManuallyFiltered(page) {
+    // Skip pages that have been manually filtered and therefore have a path like
+    // - guides-and-tutorials/index.websites.html
+    // - guides-and-tutorials/index.email.amp.html
+    // But still the matching format should be filtered, therefore pass it back
+    const format = page.path.match(/\.(websites|stories|ads|email)(\.amp)?\.html/);
+    if (format) {
+      return format[1];
+    }
   }
 
+  /**
+   * Checks if a filtered variant already for a specific format already exists
+   * on disc
+   *
+   * @param  {Vinyl}
+   * @param  {String}
+   * @return {undefined}
+   */
+  _hasManualFiltered(page, format) {
+    // Do not filter pages that have a manually filtered equivalent as they
+    // are also somewhere in the stream and shouldn't be overwritten
+    let path = page.path.replace('.amp.html', '.html');
+    path = path.replace('.html', `.${format}.html`);
+    if (fs.existsSync(path)) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Filters the given pages by the formats defined in data-available-formats
+   *
+   * @param  {Array} ...pages
+   * @return {Array}
+   */
+  _filterPages(...pages) {
+    const filteredPages = [];
+    for (const page of pages) {
+      const html = page.contents.toString();
+      const manualFormat = this._isManuallyFiltered(page);
+      if (manualFormat) {
+        const filteredHtml = this.filterHtml(html, manualFormat, true);
+        page.contents = Buffer.from(filteredHtml);
+        continue;
+      }
+
+      for (const format of FORMATS) {
+        if (this._hasManualFiltered(page, format)) {
+          continue;
+        }
+
+        const filteredHtml = this.filterHtml(html, format);
+        if (filteredHtml) {
```

Comment:
this is always true as you're now returning the input html string.

Reply:
Thanks, returns `undefined` for unfiltered HTML again now, which is already gracefully handled by both `pageTransformer.js` and `pages.js` router.